x/finality/keeper/msg_server.go

```diff
@@ -97,9 +97,10 @@ func (ms msgServer) AddFinalitySig(goCtx context.Context, req *types.MsgAddFinal
 	}
 	existingSig, err := ms.GetSig(ctx, req.BlockHeight, fpPK)
 	if err == nil && existingSig.Equals(req.FinalitySig) {
-		ms.Logger(ctx).Debug("Received duplicated finiality vote", "block height", req.BlockHeight, "finality provider", req.FpBtcPk)
-		// exactly same vote already exists, return success to the provider
-		return &types.MsgAddFinalitySigResponse{}, nil
+		ms.Logger(ctx).Debug("Received duplicated finality vote", "block height", req.BlockHeight, "finality provider", req.FpBtcPk)
+		// exactly same vote already exists, return error
+		// this is to secure the tx refunding against duplicated messages
+		return nil, types.ErrDuplicatedFinalitySig
```

Comment:
I would wait for @gitferry review here as he may have some insight how this will influence finality provider side

Reply:
I think we will need to add this error as expected error [here](https://github.com/babylonlabs-io/finality-provider/blob/74baa4c0cb1e802608a3f721064cab1dfbb99267/clientcontroller/babylon.go#L178-L259). After that it should be good

Reply:
Thanks for the waiting. I agree with @SebastianElvis that we can add the duplicate error as expected in the finality provider. I guess this applies in covenant emulator sending duplicate sigs as well. Added issue https://github.com/babylonlabs-io/finality-provider/issues/84 and https://github.com/babylonlabs-io/covenant-emulator/issues/17 to track this